src/service/layers-impl.js

```diff
@@ -0,0 +1,634 @@
+    // everything in this layer.
+    const contained = layer.contains(this.element_);
+
+    filterSplice(this.children_, layout => {
+      if (contained || layer.contains(layout.element_)) {
+        layout.parentLayer_ = layer;
+        layer.children_.push(layout);
+        return false;
+      }
+
+      return true;
+    });
+  }
+
+  /**
+   * Gets the element's parent layer. If this element is itself a layer, it
+   * returns the layer's parent.
+   * @return {?LayoutElement}
+   */
+  getParentLayer() {
+    if (this.parentLayer_ === undefined) {
+      const parent = LayoutElement.getParentLayer(this.element_, true);
+      this.parentLayer_ = parent;
+      if (parent) {
+        parent.add(this.element_);
+      }
+    }
+    return this.parentLayer_;
+  }
+
+  forgetParentLayer() {
+    this.parentLayer_ = undefined;
+  }
+
+  /**
+   * Gets the size of the element.
+   * @return {!SizeDef}
+   */
+  getSize() {
+    if (this.needsRemeasure_) {
+      this.remeasure();
+    }
+    return this.size_;
+  }
+
+  /**
+   * Gets offsets of the element relative to the parent layer, without taking
+   * scroll position into account.
+   * @return {!PositionDef}
+   */
+  getOffsetFromParent() {
+    if (this.needsRemeasure_) {
+      this.remeasure();
+    }
+    return this.position_;
+  }
+
+  getScrollTop() {
+    this.updateScrollPosition_();
+    return this.scrollTop_;
+  }
+
+  getScrollLeft() {
+    this.updateScrollPosition_();
+    return this.scrollLeft_;
+  }
+
+  /**
+   * Gets the position of the element relative to the parent layer, taking
+   * scroll position of the parents into account.
+   * @param {Element=} opt_parent
+   * @return {!LayoutRectDef}
+   */
+  getScrolledPosition(opt_parent) {
+    const position = this.getOffsetFromParent();
+    let x = position.left;
+    let y = position.top;
+
+    let last;
+    const stopAt = opt_parent ? LayoutElement.getParentLayer(opt_parent) : null;
+    for (let p = this.getParentLayer(); p !== stopAt; p = p.getParentLayer()) {
+
+      x -= p.getScrollLeft();
+      y -= p.getScrollTop();
+
+      if (last) {
+        const position = last.getOffsetFromParent();
+        x += position.left;
+        y += position.top;
+      }
+      last = p;
+    }
+
+    return PositionLt(x, y);
+  }
+
+  /**
+   * Gets the position of the element relative to the parent layer, taking
+   * scroll position of the parents into account.
+   * @param {Element=} opt_parent
+   * @return {!LayoutRectDef}
+   */
+  getOffsetPosition(opt_parent) {
+    const position = this.getOffsetFromParent();
+    let x = position.left;
+    let y = position.top;
+
+    let last;
+    const stopAt = opt_parent ? LayoutElement.getParentLayer(opt_parent) : null;
+    for (let p = this.getParentLayer(); p !== stopAt; p = p.getParentLayer()) {
+      if (last) {
+        const position = last.getOffsetFromParent();
+        x += position.left;
```

Comment:
This case is likewise confusing. But understand we simply need it for adx/ady. I'm ok with it. Just want to limit the exposure to the API that might be used in 1-2 places, but otherwise makes very little sense.

Reply:
I'm not sure how to address this comment, either.

Reply:
Let's discuss on the thread above.